### PR TITLE
Conditional Create/Update respects SearchEntryMode

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
Conditional Create/Update must respect SearchEntryMode.Match when searching for existing items

## Related issues
Addresses #730

## Testing
Existing unit/integration tests

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs [hunk 1]
```diff
@@ -69,9 +71,9 @@ public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceReques
             }
             else if (count == 1)
             {
-                ResourceWrapper resourceWrapper = results.Results.First().Resource;
-                var resource = message.Resource.ToPoco();
-                var version = WeakETag.FromVersionId(resourceWrapper.Version);
+                ResourceWrapper resourceWrapper = matchedResults.First().Resource;
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: nit: matchedResults[0]

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs [hunk 2]
```diff
@@ -50,8 +51,9 @@ public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceReques
             EnsureArg.IsNotNull(message, nameof(message));
 
             SearchResult results = await _searchService.SearchAsync(message.Resource.InstanceType, message.ConditionalParameters, cancellationToken);
+            SearchResultEntry[] matchedResults = results.Results.Where(x => x.SearchEntryMode == ValueSets.SearchEntryMode.Match).ToArray();
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: Could I do a search with a `_count=1` parameter and force the execution?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/ConditionalCreateResourceHandler.cs [hunk 3]
```diff
@@ -44,8 +43,9 @@ public async Task<UpsertResourceResponse> Handle(ConditionalCreateResourceReques
             EnsureArg.IsNotNull(message, nameof(message));
 
             SearchResult results = await _searchService.SearchAsync(message.Resource.InstanceType, message.ConditionalParameters, cancellationToken);
+            SearchResultEntry[] matchedResults = results.Results.Where(x => x.SearchEntryMode == ValueSets.SearchEntryMode.Match).ToArray();
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/ConditionalCreateResourceHandler.cs]: If I search with `_summary=count` would this force execution?

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/ConditionalCreateResourceHandler.cs]: I don't think either of these options are covered by the spec, but possibly we should filter out these parameters

Comment (brandonpollett) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/ConditionalCreateResourceHandler.cs]: Yea, my question is more around if we would execute the search with those parameters thus possibly having an unexpected outcome.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/BaseConditionalHandler.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            // Filters search parameters that can limit the number of exact results
+            IReadOnlyList<Tuple<string, string>> filteredParameters = conditionalParameters
+                .Where(x => !string.Equals(x.Item1, KnownQueryParameterNames.Count, StringComparison.OrdinalIgnoreCase)
+                            && !string.Equals(x.Item1, KnownQueryParameterNames.Summary, StringComparison.OrdinalIgnoreCase))
+                .ToArray();
+
+            SearchResult results = await _searchService.SearchAsync(instanceType, filteredParameters, cancellationToken);
+
+            // No search parameters where used
+            if (results?.UnsupportedSearchParameters.Count + results?.UnsupportedSortingParameters.Count == filteredParameters.Count)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/BaseConditionalHandler.cs]: Clean this up

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/BaseConditionalHandler.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            EnsureArg.IsNotNull(searchService, nameof(searchService));
+
+            _searchService = searchService;
+        }
+
+        protected async Task<SearchResultEntry[]> Search(string instanceType, IReadOnlyList<Tuple<string, string>> conditionalParameters, CancellationToken cancellationToken)
+        {
+            // Filters search parameters that can limit the number of exact results
+            IReadOnlyList<Tuple<string, string>> filteredParameters = conditionalParameters
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/BaseConditionalHandler.cs]: Also filter _sort

**Pull request merged**